internal/users/users_test.go

```diff
@@ -121,17 +122,174 @@ func TestUpdateLocalGroups(t *testing.T) {
+				users.WithGetUsersCmd(getUsersCmd),
+			}
+			err := users.CleanSystemGroups(cleanupOptions...)
+			if tc.wantErr {
+				require.Error(t, err, "CleanupSystemGroups should have failed")
+				return
+			}
+			require.NoError(t, err, "CleanupSystemGroups should not have failed")
+
+			// Always check the golden files missing for no-op too on error
+			referenceFilePath := testutils.GoldenPath(t)
+			if testutils.Update() {
+				// The file may already not exists.
+				_ = os.Remove(testutils.GoldenPath(t))
+				referenceFilePath = destCmdsFile
+			}
+
+			var shouldExists bool
+			if _, err := os.Stat(referenceFilePath); err == nil {
+				shouldExists = true
+			}
+			if !shouldExists {
+				require.NoFileExists(t, destCmdsFile, "UpdateLocalGroups should not call gpasswd by did")
+				return
+			}
+
+			got := usertests.IdempotentGPasswdOutput(t, destCmdsFile)
+			want := testutils.LoadWithUpdateFromGolden(t, got)
+			require.Equal(t, want, got, "Clean up should do the expected gpasswd operation, but did not")
+		})
+	}
+}
+
+func TestCleanUserFromSystemGroups(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username string
+
+		groupFilePath   string
+		wantMockFailure bool
+
+		wantErr bool
+	}{
+		"Cleans up user from group":                   {},
+		"Cleans up user from multiple groups":         {groupFilePath: "user_in_many_groups.group"},
+		"No op if user does not belong to any groups": {username: "groupless"},
+
+		"Error on missing groups file":                {groupFilePath: "does_not_exists.group", wantErr: true},
+		"Error when groups file is malformed":         {groupFilePath: "malformed_file.group", wantErr: true},
+		"Error on any unignored delete gpasswd error": {wantMockFailure: true, wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			if tc.username == "" {
+				tc.username = "myuser"
+			}
+			if tc.groupFilePath == "" {
+				tc.groupFilePath = "user_in_one_group.group"
+			}
+
+			destCmdsFile := filepath.Join(t.TempDir(), "gpasswd.output")
+			groupFilePath := filepath.Join("testdata", tc.groupFilePath)
+			gpasswdCmd := []string{"env", "GO_WANT_HELPER_PROCESS=1",
+				fmt.Sprintf("GO_WANT_HELPER_PROCESS_DEST=%s", destCmdsFile),
+				fmt.Sprintf("GO_WANT_HELPER_PROCESS_GROUPFILE=%s", groupFilePath),
+				os.Args[0], "-test.run=TestMockgpasswd", "--",
+			}
+			if tc.wantMockFailure {
+				gpasswdCmd = append(gpasswdCmd, "gpasswdfail")
+			}
+
+			cleanupOptions := []users.Option{
+				users.WithGpasswdCmd(gpasswdCmd),
+				users.WithGroupPath(groupFilePath),
+			}
+			err := users.CleanUserFromSystemGroups(tc.username, cleanupOptions...)
+			if tc.wantErr {
+				require.Error(t, err, "CleanupSystemGroups should have failed")
+				return
+			}
+			require.NoError(t, err, "CleanupSystemGroups should not have failed")
+
+			// Always check the golden files missing for no-op too on error
+			referenceFilePath := testutils.GoldenPath(t)
```

Comment:
Mind adding a Jira card that we should cleanup that (it’s now the 4th implementation to it and I think we should factor it out, get that in either idempotent or something similar)

Reply:
Oh good if you want to tackle it right away, I didn’t want to delay this PR more hence the possibility to have it separated, but your pick!

Reply:
Following the discussion that we had during the standup, this will require more refactoring and thinking to do it properly, so it's better to only create the card and tackle it later.

The card is already created!